client/base_client.go

```diff
@@ -12,3 +13,31 @@ type BaseClient interface {
 	SendRequest(method string, rawURL string, data url.Values,
 		headers map[string]interface{}) (*http.Response, error)
 }
+
+// BaseClientWithCtx is an extension of BaseClient with the ability to associate a contex with
+// the request
+type BaseClientWithCtx interface {
+	BaseClient
+	SendRequestWithCtx(ctx context.Context, method string, rawURL string, data url.Values,
+		headers map[string]interface{}) (*http.Response, error)
+}
+
+type wrapperClient struct {
+	// embed the BaseClient so the functions remain accessible
+	BaseClient
+}
+
+func (w wrapperClient) SendRequestWithCtx(ctx context.Context, method string, rawURL string, data url.Values,
+	headers map[string]interface{}) (*http.Response, error) {
+	return w.SendRequest(method, rawURL, data, headers)
```

Comment:
Should this be calling `SendRequestWithCtx` instead? Otherwise the context is swallowed.

Reply:
Sorry, there really should've been a comment over `wrapperClient` to make this more explicit.


`wrapperClient` is wrapping the older `BaseClient` implementation. So this method is the new interface passing though to the context-free client.
This enables the SDK to use the `BaseClientWithCtx` interface everywhere since we can transform the older `BaseClient`

Reply:
Ah, understanding better now the intent. I'm thinking it would be a little easier to understand the implications by actually having the `wrapperClient` send through the provided context (which is what I was expecting in the initial review).

wrapping/upgrading is adding the ability to an existing client/handler to pass a context through (by adding an implementation for `SendRequestWithCtx` to it). So it no longer becomes a noop that swallows the context. That way if I wanted to take an existing integration and start adding context to API calls, all I'd need to do is switch to use the context-based version of the API method (no initialization changes needed to the client/service/handler).

Thoughts?

Reply:
So there are a few different ways for folks to instantiate a twiliogo client, in all but one the `*WithCtx` methods will operate as expected. There's just one scenario where it will be misleading to the integration developer, but I think it's the most sophisticated setup so there's the highest likelihood they will identify this.

Integrations where the client will "just work" with `*WithCtx` methods:
- generic [`twiliogo.NewRestClient`](https://pkg.go.dev/github.com/twilio/twilio-go#NewRestClient) is upgraded automatically
- [`client.Client`](https://pkg.go.dev/github.com/twilio/twilio-go/client#Client) that provides a `*http.Client` will automatically use the context

The only scenario where a context may be provided but not used is when the integration is using [`client.RequestHandler`](https://pkg.go.dev/github.com/twilio/twilio-go/client#RequestHandler) and has provided a custom `BaseClient`. If this is the case, there's no way for the library to pass the context in unless we introduce breaking changes, since the specified client by the integration (which is opaque to the SDK) has no method available to ingest a context.

Reply:
Ah, now I I understand. Thanks for bearing with me.